### ISSUE TITLE
Interact with julia scripts a bit like "weave.jl

### ISSUE THREAD

**Issue by tlienart** [tlienart/JuDoc.jl#103]
A possible solution is to have something like 

```md
Some markdown here

\include{aux/script1.jl}    <!-- writes the script content -->

\include{aux/out/script1.txt} <!-- writes the results (if text) -->

![]{aux/out/script1_img1.png}    <!-- or show an image generated by the script -->
```

`script1.jl` is then just an auxiliary file that contains some julia code that possibly prints stuff (and the stuff would go to `script1.txt`, see below) or saves some image(s). 

The advantage with this is multiple

1. it's easy to implement (just requires to add the `include` command)
2. the scriptX.jl are directly runnable, so can check that everything is fine outside of JuDoc, also it doesn't slow JuDoc down because it would be done separately
3. the scriptX.jl can be called by a big file `generate_results.jl` which effectively acts as a big "test" making sure that all the scripts run fine, again this is outside of JuDoc.

## A simple example

```md
# Title
## An experiment

You can run this code

\include{aux/script1.jl}

and you will get this result

\include{aux/out/script1_res.txt}

also this plotting code

\include{aux/script2.jl}

will show you this image

![](aux/out/script1_img1.png)

etc...
```

the file `script1.jl`

```julia
for i in 1:5
  println("*"^i)
end
```

the file `script2.jl`

```julia
using PyPlot # HIDE
plot(randn(5))
savefig(...) # HIDE
```

the file `generate_results.jl`

```julia
# this file would be a bit like runtests.jl
# here potentially load libraries that are needed

# whatever scripts print would go to the `_res` file.
for script in ["script1.jl", ]
    open(joinpath("aux", "out", "$(script)_res.txt"), "w") do io
        redirect_stdout(io) do
            include(joinpath("aux", script))
        end
    end
    # here could check if the `res` file is empty, and if so remove it.
end
```

**Edit**: I'll demo how that can work in coming few days in the JudocExample repo.

**Comment by tlienart** [comment]
Note that it would be possible to have the code directly in the markdown, something like 

```julia
# Title
# an experiment

```julia:EXECUTETHIS
for i in 1:5
  println("*"^i)
end
```
```

and have JuDoc execute the code and display the results in its place but I think this leads to several issues:

1. it requires more thinking to avoid re-running the code every time you fix the webpage, a possibility is to build a dictionary of all the code snippets and only run them when they change
2. (in my opinion a big issue) you can't directly test your code, e.g. if you're in Juno, you can't execute that bit of code, you'd have to execute JuDoc, and if it fails on your code, it would also crash the current visualisation of the website which is not great, you'd have to restart with `serve()`. 

So long story short I think the solution above is potentially easier to implement and not too bad for the user either?

**Comment by asprionj** [comment]
I think we have to decide where JuDoc fits in. There's [Documenter.jl](https://juliadocs.github.io/Documenter.jl/stable/index.html) for heavily code-oriented writing, e.g. a collection of use cases, examples, code snippets, etc. With some work, this could probably be extended towards a homepage-like static site. There's [Weave.jl](http://weavejl.mpastell.com/stable/) for creating a "documented script" (or "scientific report", as they call it). And, of course, the well-known ones like Sphinx or MkDocs.

There's also online-markdown-to-site-converters; [docsify](https://docsify.js.org/) comes to my mind. It offers live previewing before packaging everything in a bundle of HTML, CSS and JS that can be statically hosted (on GitHub pages, for example).

Then there's all those CMS's. I think [Grav](https://getgrav.org/) is an awesome one. Simple, yet flexible and powerful. Check, for example, all the [templates](https://getgrav.org/downloads/themes) available! Finally, there's the myriad of static site generators.

So, what do we need, and for whom? Personally, I'd like to be able to just type some sort of simple markdown, select from some nice but simple themes / (page/site) styles and layouts, directly write LaTeX inside the markdown files, and include code examples (in Julia) for which the output and plots are automatically included, if I wish so (in SVG format). This all with a nice live preview and the possibility of directly deploying everything as a static site. I feel like JuDoc is already quite close to this... your thoughts?

---

Concerning inline code versus external .jl files: I agree that separating code and markdown is a good thing. Then one can use / call / share / ... the code also separated from the page / markdown. Also, this simplifies deciding whether and which code sample has changed. However, I feel like it would be a major benefit if the samples were re-generated automatically upon changes. Of course, we'd have to keep track. And of course, we'd have to keep an instance of Julia running in the background, with everything loaded from the previous / initial rendering pass -- to make re-rendering fast.

---

A side note: The notion of "slimming down" the pages should be extended to code highlighting. Why ship the code-highlighting runtime and the non-highlighted code to the client, when the highlighting could be done once at page-creation time?

**Comment by tlienart** [comment]
Yes I agree with all this, in my opinion JuDoc would fit in the "personal/small website" box a-la Hugo, Jekyll or indeed Grav. 

Anyway with respect to code it seems like the solution above to show code would work well, and then in the longer run it could be good to have something that does some sort of tracking so that the displayed output is always certified to be in sync with the code (and I think you're maybe suggesting threads or processes here and that would indeed definitely be a possibility). I'll try to do a basic implementation and show how it could work. 

------------------------------------

You also mention two other elements with which I agree:

* having a few good CSS templates could really help
* do the code highlighting in JuDoc

The first one I think is straightforward and just requires a bit of work, possibly some of the templates could just be copied and adapted from open ones that are available. 
The second point is potentially a bit more tricky but I think it's a very nice idea. Issue https://github.com/tlienart/JuDoc.jl/issues/104 can track discussion on that point.

thanks for your input!

**Comment by tlienart** [comment]
Ok I've added a prototype for the feature, it's included in the JuDocExample repo, this part:

```
## Including scripts

In order to include the content of a script you can use the `\input` command specifying which language it should be tagged as:

\input{julia}{scripts/script1.jl}

these scripts can be run in such a way that their output is also saved to file, see `scripts/generate_results.jl` and you can then also input the results:

\input{}{scripts/output/script1_res.txt}

which is convenient if you're presenting code. The `generate_results.jl` file also makes sure that all the code on your website works and that all results match the code which makes maintenance easy.
```

renders to

![screen shot 2019-03-05 at 2 14 06 pm](https://user-images.githubusercontent.com/10897531/53778507-147abe00-3f51-11e9-9387-a0cf2137ee64.png)

What do you think @asprionj ?

The remaining thing to explore is to track changes, I'll think about it.

**Comment by asprionj** [comment]
Looks good.

Does one have to write his own `generate_results.jl` file, and also run it manually? Or could we auto-generate it from extracting all the `\include` statements? Probably we'd have to have an additional commodity feature to ensure consistency... let me explain:

Now, the user includes code, output and possibly plots manually. So the .txt or graphic files could be anything and don't directly relate to any code. How about something like:
```
Include the code:
\input{code:julia}{scripts/script1.jl}

Include the output:
\input{output}{scripts/script1.jl}

And possibly some plots, here the second and fourth one produces by the script:
\input{plot:2,4}{scripts/script1.jl}
``` 

The parser would then see that all reference to the same code file. And all the handling of auxiliary files would be automatic and hidden. Is Julia able of auto-capturing all plots? I.e. redirect them to files?

**Comment by tlienart** [comment]
Ok so the API you're describing would just be a small modification from what I've done so that's easy.

The generate results (maybe it should be called `run_scripts.jl`) would have to be written by the user for now yes, it's not directly clear to me how to do that automatically without it causing problems.

However it could be good to have a simple macro like `@gen "script1.jl"` so that the file would look clean and more like a test file: 

```
using JuDoc:@gen

@gen "script1.jl"
@gen "script2.jl"
... 
```

Also to make things more integrated, the `serve` command could take a keyword like `serve(run_scripts=true)` which would then start by running all scripts before the website is served.

**Comment by tlienart** [comment]
Update:

* `\include{julia}{script.jl}`, `\include{code:julia}{script.jl}`
* `\include{output}{script.jl}`, `\include{output:plain}{script.jl}`
* `\include{plot}{script.jl}`, `\include{plot:id}{script.jl}`

now all work (cf also the example repo). A few things to note:

1. the `output:plain` was to have consistency with the rest but could also lead to a `output:table` to display something that comes out of a DataFrames for instance. I'm not going to push this too much now as I think that's getting close to Weave territory while there is still a lot to do away from that 🙂 
1. I decided against `plot:1,4` because I felt that the most relevant use case would be to display plots one at the time and possibly say something between plots. It's still possible to do `\include{plot:1}{script.jl} \include{plot:2}{script.jl} \include{plot:3}{script.jl}` if necessary.

This probably needs some more testing but I guess it's a nice new feature, thanks for the suggestions. I'll close this for now but I suspect related issues / feature requests might pop up after a bit more testing
